libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/DialogManager.cs

```diff
@@ -227,21 +224,126 @@ public async Task<DialogManagerResult> OnTurnAsync(ITurnContext context, Cancell
+        {
+            // the bot is running as a skill. 
+            var turnContext = dc.Context;
+
+            // Process remote cancellation
+            if (turnContext.Activity.Type == ActivityTypes.EndOfConversation && dc.ActiveDialog != null && IsEocComingFromParent(turnContext))
+            {
+                // Handle remote cancellation request from parent.
+                var activeDialogContext = GetActiveDialogContext(dc);
+
+                var remoteCancelText = "Skill was canceled through an EndOfConversation activity from the parent.";
+                await turnContext.TraceActivityAsync($"{typeof(Dialog).Name}.RunAsync()", label: $"{remoteCancelText}", cancellationToken: cancellationToken).ConfigureAwait(false);
+
+                // Send cancellation message to the top dialog in the stack to ensure all the parents are canceled in the right order. 
+                return await activeDialogContext.CancelAllDialogsAsync(true, cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+
+            // Handle reprompt
+            // Process a reprompt event sent from the parent.
+            if (turnContext.Activity.Type == ActivityTypes.Event && turnContext.Activity.Name == DialogEvents.RepromptDialog)
             {
-                snapshotDc = snapshotDc.Child;
+                await dc.RepromptDialogAsync(cancellationToken).ConfigureAwait(false);
+
+                // TODO: check with steve what we return here.
+                return new DialogTurnResult(DialogTurnStatus.Waiting);
+            }
+
+            // Continue execution
+            // - This will apply any queued up interruptions and execute the current/next step(s).
+            var turnResult = await dc.ContinueDialogAsync(cancellationToken).ConfigureAwait(false);
+            if (turnResult.Status == DialogTurnStatus.Empty)
+            {
+                // restart root dialog
+                var startMessageText = $"Starting {_rootDialogId}.";
+                await turnContext.TraceActivityAsync($"{typeof(Dialog).Name}.RunAsync()", label: $"{startMessageText}", cancellationToken: cancellationToken).ConfigureAwait(false);
+                turnResult = await dc.BeginDialogAsync(_rootDialogId, cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+
+            await SendStateSnapshotTraceAsync(dc, "Skill State", cancellationToken).ConfigureAwait(false);
+
+            // Send end of conversation if it is completed or cancelled.
+            if (turnResult.Status == DialogTurnStatus.Complete || turnResult.Status == DialogTurnStatus.Cancelled)
+            {
+                var endMessageText = $"Dialog {_rootDialogId} has **completed**. Sending EndOfConversation.";
+                await turnContext.TraceActivityAsync($"{typeof(Dialog).Name}.RunAsync()", label: $"{endMessageText}", value: turnResult.Result, cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
What if they've already sent an EndOfConversation?  I would setup a listner to detect that they sent an EOC and only send one yourself if they haven't

Reply:
Do you have any samples on how to setup a listener? I am not familiar with that.

Reply:
Hi @stevenic, I've been thinking about this, in what cases a dev would send an EoC by hand? Wouldn't that be a bug in their code?